FreeRTOS_ARP.c

```diff
@@ -119,27 +145,102 @@ eFrameProcessingResult_t eARPProcessPacket( ARPPacket_t * const pxARPFrame )
+        {
+            /* We have waited long enough, reset the counter. */
+            uxARPClashCounter = 0;
+        }
+    }
+
+    /* Introduce a do while loop to allow use of breaks. */
+    do
+    {
+        /* Only Ethernet hardware type is supported.
+         * Only IPv4 address can be present in the ARP packet.
+         * The hardware length (the MAC address) must be 6 bytes. And,
+         * The Protocol address length must be 4 bytes as it is IPv4. */
+        if( ( pxARPHeader->usHardwareType != ipARP_HARDWARE_TYPE_ETHERNET ) ||
+            ( pxARPHeader->usProtocolType != ipARP_PROTOCOL_TYPE ) ||
+            ( pxARPHeader->ucHardwareAddressLength != ipMAC_ADDRESS_LENGTH_BYTES ) ||
+            ( pxARPHeader->ucProtocolAddressLength != ipIP_ADDRESS_LENGTH_BYTES ) )
+        {
+            /* One or more fields are not valid. */
+            iptraceDROPPED_INVALID_ARP_PACKET( pxARPHeader );
+            break;
+        }
+
+        /* Check whether the lowest bit of the highest byte is 1 to check for
+         * multicast address or even a broadcast address (FF:FF:FF:FF:FF:FF). */
+        if( ( pxARPHeader->xSenderHardwareAddress.ucBytes[ 0 ] & 0x01U ) == 0x01U )
+        {
+            /* Senders address is a multicast OR broadcast address which is not
+             * allowed for an ARP packet. Drop the packet. See RFC 1812 section
+             * 3.3.2. */
+            iptraceDROPPED_INVALID_ARP_PACKET( pxARPHeader );
+            break;
+        }
+
+        uint32_t ulHostEndianProtocolAddr = FreeRTOS_ntohl( ulSenderProtocolAddress );
+
+        if( ( ipFIRST_LOOPBACK_IPv4 <= ulHostEndianProtocolAddr ) &&
+            ( ulHostEndianProtocolAddr < ipLAST_LOOPBACK_IPv4 ) )
+        {
+            /* The local loopback addresses must never appear outside a host. See RFC 1122
+             * section 3.2.1.3. */
+            iptraceDROPPED_INVALID_ARP_PACKET( pxARPHeader );
+            break;
+        }
+
+        /* Check whether there is a clash with another device for this IP address. */
+        if( ( ulSenderProtocolAddress == *ipLOCAL_IP_ADDRESS_POINTER ) &&
+            ( *ipLOCAL_IP_ADDRESS_POINTER != 0UL ) )
+        {
+            if( uxARPClashCounter < arpIP_CLASH_MAX_RETRIES )
+            {
+                /* Increment the counter. */
+                uxARPClashCounter++;
+
+                /* Send out a defensive ARP request. */
+                FreeRTOS_OutputARPRequest( *ipLOCAL_IP_ADDRESS_POINTER );
+
+                /* Since an ARP Request for this IP was just sent, do not send a gratuitous
+                 * ARP for arpGRATUITOUS_ARP_PERIOD. */
+                xLastGratuitousARPTime = xTaskGetTickCount();
+
+                /* Note the time at which this request was sent. */
+                vTaskSetTimeOutState( &xARPClashTimeOut );
+
+                /* Reset the time-out period to the given value. */
+                xARPClashTimeoutPeriod = pdMS_TO_TICKS( arpIP_CLASH_RESET_TIMEOUT_MS );
```

Comment:
Shouldn't `xARPClashTimeOut` be set to `0` so that its members are also reset?

Reply:
Not actually, it should be set to the `arpIP_CLASH_RESET_TIMEOUT_MS` value so that up in line 164, the value can be checked. Whenever the value is checked in `xTaskCheckForTimeOut`, the value of `xARPClashTimeoutPeriod` gets changed (reduced).
What reset means in this context is to set the value back to its original value.